File: python/spider_browser/ai/act.py
"""act() — single action from natural language."""

from __future__ import annotations

import asyncio

from ..protocol.protocol_adapter import ProtocolAdapter
from .llm_provider import LLMProvider
from .prompts import SYSTEM_PROMPT, build_user_message
from .agent import execute_action


async def act(
    adapter: ProtocolAdapter,
    llm: LLMProvider,
    instruction: str,
) -> None:
    """
    Execute a single action from natural language.

    Takes a screenshot + HTML, sends to LLM with the instruction,
    then executes the returned action steps.
    """
    screenshot, html, url, title = await asyncio.gather(
        adapter.capture_screenshot(),
        adapter.get_html(),
        adapter.evaluate("window.location.href"),
        adapter.evaluate("document.title"),
    )

    context = f"Task: {instruction}\nPAGE TITLE: {title}"

    plan = await llm.chat_json([
        {"role": "system", "content": SYSTEM_PROMPT},
        {"role": "user", "content": build_user_message(str(url), str(html), str(screenshot), context)},
    ])

    steps = plan.get("steps", []) if isinstance(plan, dict) else []
    for step in steps:
        await execute_action(adapter, step)
        await asyncio.sleep(0.2)

File: go/act.go
package spiderbrowser

import (
	"fmt"
	"time"
)

// Act executes a single action from natural language.
func Act(adapter *ProtocolAdapter, llm LLMProvider, instruction string) error {
	screenshot, err := adapter.CaptureScreenshot()
	if err != nil {
		return err
	}
	html, err := adapter.GetHTML()
	if err != nil {
		return err
	}
	urlVal, _ := adapter.Evaluate("window.location.href")
	titleVal, _ := adapter.Evaluate("document.title")
	url, _ := urlVal.(string)
	title, _ := titleVal.(string)

	context := fmt.Sprintf("Task: %s\nPAGE TITLE: %s", instruction, title)

	var plan AgentPlan
	err = llm.ChatJSON([]LLMMessage{
		{Role: "system", Content: SystemPrompt},
		{Role: "user", Content: BuildUserMessage(url, html, screenshot, context)},
	}, &plan)
	if err != nil {
		return err
	}

	for _, step := range plan.Steps {
		if err := ExecuteAction(adapter, step); err != nil {
			return err
		}
		time.Sleep(200 * time.Millisecond)
	}
	return nil
}
